Enable fast gradcheck for real inputs and outputs

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55871 Refactor fast gradcheck
* #55699 Enable fixed fast_mode for complex
* #55692 Refactor C->C to C->R twice
* #55656 Clean up raise exception logic
* **#55237 Enable fast gradcheck for real inputs and outputs**
* #56279 Disable flaky tests

In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Bolded indicates the issue is handled in this PR, otherwise it is handled in a previous PR.

**Non-determinism issues**:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)

Randomness:
  - RRelu (new module tests) - we fix by using our own generator as to avoid messing with user RNG state (handled in #54480)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally) - we fixed this by disabling for float32 (handled in previous  PR)
- cholesky_solve (test_linalg): #56235 handled in previous PR
- **cumprod** (test_ops) - #56275 disabled fast gradcheck

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D27920906](https://our.internmc.facebook.com/intern/diff/D27920906)

soulitzer committed:
Enable fast gradcheck for real inputs and outputs

[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"




[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)
- cases where setting torch.manual_seed fixes
  - RRelu (new module tests)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)

Randomness:
  - RRelu (new module tests) - we fix this one by using our own generator as to avoid messing with user RNG state

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally),
- cholesky_solve (test_linalg): https://app.circleci.com/pipelines/github/pytorch/pytorch/296020/workflows/27142826-4629-44e9-a365-b412aec374c5/jobs/12113886
- cumprod (test_ops)

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Non-determinism issues:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)

Randomness:
  - RRelu (new module tests) - we fix by using our own generator as to avoid messing with user RNG state

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally) - we fixed this by disabling for float32 (handled in previous  PR)
- cholesky_solve (test_linalg): #56235 handled in previous PR
- cumprod (test_ops) - #56275 

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

albanD commented on test/test_ops.py:
```diff
@@ -6,13 +6,13 @@
 from torch.testing._internal.common_device_type import \
     (instantiate_device_type_tests, ops, onlyCPU, onlyOnCPUAndCUDA, skipCUDAIfRocm, OpDTypes)
 from torch.testing._internal.common_jit import JitCommonTestCase, check_against_reference
-from torch.autograd.gradcheck import gradcheck, gradgradcheck
```
cc @mruberry  FYI, we're changing here to use the test-specific version of gradcheck so that it will use the proper default for fast vs slow gradcheck.
It will also start using the other defaults about batched grad but the argument is always passed and so not changing anything.

Is that ok to do? Or you prefer to keep the plain gradcheck here with special logic to handle default value?

mruberry replied:
Seems OK to me

albanD reviewed (commented):
The update looks good to me.
I just want to confirm with @mruberry  that the changes in OpInfo match the design idea there.

albanD reviewed (approved):
Good to go!

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Bolded indicates the issue is handled in this PR, otherwise it is handled in a previous PR.

**Non-determinism issues**:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)

Randomness:
  - RRelu (new module tests) - we fix by using our own generator as to avoid messing with user RNG state (handled in #54480)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally) - we fixed this by disabling for float32 (handled in previous  PR)
- cholesky_solve (test_linalg): #56235 handled in previous PR
- **cumprod** (test_ops) - #56275 disabled fast gradcheck

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)


[ghstack-poisoned]

soulitzer committed:
Update on "Enable fast gradcheck for real inputs and outputs"


In this PR, we reenable fast-gradcheck and resolve misc issues that arise:
Before landing this PR, land #55182 so that slow tests are still being run periodically.

Bolded indicates the issue is handled in this PR, otherwise it is handled in a previous PR.

**Non-determinism issues**:
- ops that do not have deterministic implementation (as documented https://pytorch.org/docs/stable/generated/torch.use_deterministic_algorithms.html#torch.use_deterministic_algorithms)
  - test_pad_cuda (replication_pad2d) (test_nn)
  - interpolate (test_nn)
  - cummin, cummax (scatter_add_cuda_kernel) (test_ops)
  - test_fn_gradgrad_prod_cpu_float64 (test_ops)

Randomness:
  - RRelu (new module tests) - we fix by using our own generator as to avoid messing with user RNG state (handled in #54480)

Numerical precision issues:
- jacobian mismatch: test_gelu (test_nn, float32, not able to replicate locally) - we fixed this by disabling for float32 (handled in previous  PR)
- cholesky_solve (test_linalg): #56235 handled in previous PR
- **cumprod** (test_ops) - #56275 disabled fast gradcheck

Not yet replicated:
 - test_relaxed_one_hot_categorical_2d (test_distributions)

Differential Revision: [](https://our.internmc.facebook.com/intern/diff/)

Differential Revision: [D27920906](https://our.internmc.facebook.com/intern/diff/D27920906)

[ghstack-poisoned]
